File: src/main/java/in/costea/wiles/services/TokenTransmitter.java
package in.costea.wiles.services;

import in.costea.wiles.builders.ExpectParamsBuilder;
import in.costea.wiles.data.Token;
import in.costea.wiles.data.TokenLocation;
import in.costea.wiles.exceptions.TokenExpectedException;
import in.costea.wiles.exceptions.UnexpectedEndException;
import in.costea.wiles.exceptions.UnexpectedTokenException;
import org.jetbrains.annotations.NotNull;

import java.util.LinkedList;
import java.util.List;
import java.util.Optional;
import java.util.function.Predicate;

import static in.costea.wiles.builders.ExpectParamsBuilder.*;
import static in.costea.wiles.statics.Constants.*;

public class TokenTransmitter
{
    private final LinkedList<Token> tokens;
    private final TokenLocation lastLocation;

    public TokenTransmitter(@NotNull List<Token> tokens)
    {
        this.tokens = new LinkedList<>(tokens);
        if (tokens.size() > 0)
            lastLocation = tokens.get(tokens.size() - 1).location();
        else lastLocation = null;
    }

    private void removeToken() {
        if (tokensExhausted())
            throw new IllegalStateException("Tokens exhausted!");
        tokens.pop();
    }

    public boolean tokensExhausted()
    {
        return tokens.isEmpty();
    }

    public Token expect(ExpectParamsBuilder params) throws UnexpectedEndException, UnexpectedTokenException, TokenExpectedException {
        String message = params.getErrorMessage();
        boolean succeeded = false;
        if(params.getWhenRemoveToken()==DEFAULT)
            params.removeTokenWhen(ALWAYS);
        try
        {
            if(tokensExhausted())
                throw new UnexpectedEndException(message,lastLocation);
            Token token;

            boolean shouldIgnoreNewLine = params.isIgnoringNewLine();
            if(shouldIgnoreNewLine)
            {
                while ((token=tokens.getFirst()).content().equals(NEWLINE_ID))
                {
                    if (tokensExhausted())
                        throw new UnexpectedEndException(message, token.location());
                    removeToken();
                }
            }

            token=tokens.getFirst();
            Predicate<String> foundTest= params.getFoundTest();
            if (!foundTest.test(token.content()))
                throw new TokenExpectedException(message, token.location());

            succeeded=true;
            return token;
        }
        finally
        {
            var whenRemoveToken = params.getWhenRemoveToken();
            if((!succeeded && whenRemoveToken == ALWAYS) || (succeeded && whenRemoveToken != NEVER))
                if(!tokensExhausted())
                    removeToken();
        }
    }

    public Optional<Token> expectMaybe(ExpectParamsBuilder expectParamsBuilder) throws UnexpectedTokenException
    {
        try
        {
            if(expectParamsBuilder.getWhenRemoveToken()==DEFAULT)
                expectParamsBuilder.removeTokenWhen(WHEN_FOUND);
            return Optional.of(expect(expectParamsBuilder));
        }
        catch (TokenExpectedException | UnexpectedEndException ex)
        {
            return Optional.empty();
        }
    }
}

File: src/main/java/in/costea/wiles/services/TokenTransmitter.kt
package `in`.costea.wiles.services

import `in`.costea.wiles.builders.ExpectParamsBuilder
import `in`.costea.wiles.data.Token
import `in`.costea.wiles.enums.WhenRemoveToken
import `in`.costea.wiles.exceptions.TokenExpectedException
import `in`.costea.wiles.exceptions.UnexpectedEndException
import `in`.costea.wiles.constants.Tokens
import `in`.costea.wiles.constants.ErrorMessages.INTERNAL_ERROR
import java.util.*

class TokenTransmitter(tokens: List<Token>) {
    private val tokens = LinkedList(tokens)

    @Throws(UnexpectedEndException::class, TokenExpectedException::class)
    fun expect(params: ExpectParamsBuilder): Token {
        val message = params.errorMessage
        var succeeded = false
        if (params.whenRemove == WhenRemoveToken.Default)
            params.removeWhen(WhenRemoveToken.Always)
        message!!
        return try {
            if (tokens.isEmpty()) throw UnexpectedEndException(message, null)
            if (params.isIgnoringNewLine) {
                while (tokens.first.content == Tokens.NEWLINE_ID) {
                    val token = tokens.pop()
                    if (tokens.isEmpty()) throw UnexpectedEndException(message, token.location)
                }
            }
            val foundTest = params.foundTest
            if (!foundTest.test(tokens.first.content))
                throw TokenExpectedException(message, tokens.first.location)

            succeeded = true
            tokens.first
        } finally {
            val whenRemoveToken: WhenRemoveToken = params.whenRemove
            assert(whenRemoveToken != WhenRemoveToken.Default)
            if ((!succeeded && whenRemoveToken == WhenRemoveToken.Always) || (succeeded && whenRemoveToken != WhenRemoveToken.Never))
                if (!tokens.isEmpty())
                    tokens.pop()
        }
    }

    fun expectMaybe(expectParamsBuilder: ExpectParamsBuilder): Optional<Token> {
        return try {
            if (expectParamsBuilder.whenRemove == WhenRemoveToken.Default)
                expectParamsBuilder.removeWhen(WhenRemoveToken.WhenFound)
            if (expectParamsBuilder.errorMessage == null)
                expectParamsBuilder.withErrorMessage(INTERNAL_ERROR)
            Optional.of(expect(expectParamsBuilder))
        } catch (ex: TokenExpectedException) {
            Optional.empty()
        } catch (ex: UnexpectedEndException) {
            Optional.empty()
        }
    }

    fun tokensExhausted(): Boolean {
        return tokens.isEmpty()
    }
}